keps/sig-storage/3294-provision-volumes-from-cross-namespace-snapshots/README.md

```diff
@@ -0,0 +1,1105 @@
+    A controller won't use the `VolumeSnapshot` until the `ReferenceGrant` that allows the access is created.
+
+2. `ReferenceGrant` for a `VolumeSnapshotLink` exists when the `VolumeSnapshotLink` is handled, and the `ReferenceGrant` is deleted later
+    A controller will use the `VolumeSnapshot` if the `ReferenceGrant` that allows the access exists when it checks.
+    If all the processes succeed without any error, it succeeds even the `ReferenceGrant` is deleted in the middle of the processes.
+    If any errors happened in the processes and the controller retries, it may detect that there is no `ReferenceGrant` . Then, it won't use the `VolumeSnapshot` until the `ReferenceGrant` that allows the access is re-created.
+
+#### Conflict on installing `VolumePopulator` CR for `VolumeSnapshotLink` across CSI drivers
+
+This feature requires installing `VolumePopulator` CR for `VolumeSnapshotLink` and is enabled per CSI driver basis.
+Therefore, on enabling this feature for each CSI driver, it is expected that `VolumePopulator` CR for `VolumeSnapshotLink` is created before each CSI driver installation.
+As a result, there may be a conflict in creating it for each driver, if there are any differences in their definitions, like alpha API and beta API.
+
+To avoid this issue, it should be avoided to manage `VolumePopulator` CR for `VolumeSnapshotLink` in each CSI driver's repository.
+It should be managed in another single repository and the same CR should be used per cluster basis.
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+### Example flow of how this proposal works
+
+Let's use [Story 1](#story-1) as an example and let's assume the following:
+- There are two namespaces, prod and test,
+- Alice manages the prod namespace and Bob manages the test namespace,
+- Alice would like to allow `VolumeSnapshot` foo-backup in the prod namespace to be consumed in the test namespace for testing,
+- Bob would like to create a PV for PVC foo-testing in the test namespace from the `VolumeSnapshot` foo-backup in the prod namespace.
+
+Once this proposal is implemented, it can be achieved by doing the following steps:
+
+1. In the prod namespace, Alice creates a `ReferenceGrant` bar that allows referencing to the `VolumeSnapshot` foo-backup in the prod namespace from any `VolumeSnapshotLinks` in the test namespace,
+    ```yaml
+    apiVersion: gateway.networking.k8s.io/v1alpha2
+    kind: ReferenceGrant
+    metadata:
+      name: bar
+      namespace: prod
+    spec:
+      from:
+      - group: snapshot.storage.k8s.io/v1alpha1
+        kind: VolumeSnapshotLink
```

Comment:
Have we considered an approach where we don't have a SnapshotLink object and instead we grant the PVC direct access to the Snapshot object?

Reply:
Do you mean Step 3 becomes something like this?

```
   apiVersion: v1
    kind: PersistentVolumeClaim
    metadata:
      name: foo-testing
      namespace: test
    spec:
      accessModes:
      - ReadWriteOnce
      resources:
        requests:
          storage: 10Mi
      dataSourceRef:
        apiGroup: snapshot.storage.k8s.io/v1
        kind: VolumeSnapshot
        name: foo-backup
        namespace: prod
      volumeMode: Filesystem
```

If this is what you meant, the problem is that the DataSourceRef field in PVC is TypedLocalObjectReference which does not have a namespace field.  We used TypedLocalObjectReference instead of ObjectReference to ensure PVC and datasource are in the same namespace.

Reply:
@msau42 Added description on core API change to allow specifying namespace in "Alternatives" section.